### DISCUSSION TITLE
Rename file after upload

### DISCUSSION THREAD

**Discussion by CaptainLoop** [drakkan/sftpgo#1459] [Q&A]
I am trying to create an event for renaming the file after it uploads. I want to include the timestamp, but I can't get it right. How can I do it?

I created an event action: 
**Type:** Filesystem
**Fs Action:** Rename 
Action.
**Source Path:** /data/{{Name}}/UPLOAD/{{VirtualPath}}
**Destination Path** /data/{{Name}}/UPLOAD/{{VirtualPath}}-{{Timestamp}}

and an event Rule, that includes a user test123.

In the logs, I got this message:

`Nov 16 17:41:38 serversftp sftpgo[115599]: {"level":"debug","time":"2023-11-16T17:41:38.629","sender":"HTTP","connection_id":"HTTP_clb4cgi3235000ps7350","message":"requested modification time 2023-06-22 23:48:02.816 +0200 CEST for file \"/UPLOAD/testfile.txt\", error: permission denied"}`

`Nov 16 17:41:38 serversftp sftpgo[115599]: {"level":"error","time":"2023-11-16T17:41:38.630","sender":"eventmanager","message":"unable to execute action \"renameWithTimestamp\" for rule \"Rename File with TimeStamp\", elapsed 1.503387ms, err: action \"renameWithTimestamp\" failed: fs rename failed for users: test123"}`

**Comment by sidescraper** [comment]
Same problem. I'm using Ubuntu 24.04 with Ubuntu as a PPA package.
The upload event works. > However, rename doesn't. I need that.
THX

**Comment by vlevieux** [comment]
In your `Rename` event action, try to use `{{VirtualTargetPath}}` and `{{VirtualTargetDirPath}}` instead of `{{VirtualPath}}` and `{{VirtualDirPath}}` in **both** source and destination.

Your event rule must only contain `rename` in _Fs events_, as `upload` would fail with `{{VirtualTargetPath}}`.

**Comment by sidescraper** [comment]
@vlevieux Wow! Exactly, that's what I was looking for. Just the {{TargetName}} and it's done.

**Comment by derekrhiggins** [comment]
Just for anyone coming after because I found this via google search and the comments didnt fully solve my problem - a problem I had worked out the answer to a month before but couldn't remember the solution such is the way

To do a copy or rename action the variables are like so, in my case I wanted to copy and add a .complete to the end

Action.
Source Path: {{VirtualPath}}
Destination Path: {{VirtualPath}}.complete


Hope that helps someone because there aren't enough docs/examples on this.
